### ISSUE TITLE
Distributed data collectors

### ISSUE THREAD

**Issue by vmoens** [facebookresearch/rl#141]
### Current state of data collectors
We currently have parallel data collectors, that can handle a single or multiple environments running serially or in parallel.
Say one has access to N cpus to run an experiment. She has 3 different choices:
- Create a single collector and N envs in a ParallelEnv wrapper. A rollout then looks like
```python
for t in range(n_steps):
    tensordict = policy(tensordict)  # policy is run over the previous N outputs
    tensordict = env.step(tensordict)  # there are N envs in env
    out.append(tensordict)
return out
```
- Create X = N // Y parallel envs on Y collectors. Each of the Y collectors runs its own loop:
```python
for t in range(n_steps):
    tensordict = policy(tensordict)  # policy is run over the previous X outputs
    tensordict = env.step(tensordict)  # there are X envs in env
    out.append(tensordict)
return out
```
The advantage is that we can run things asynchroneously, and get the results of the collector that finished first, while still benefiting from `ParallelEnv` and running the policy on CUDA over more than one input.
- Create a set of N collectors, each containing a single environment.

### Feature description
We'd like to be able to create distributed data collectors over multiple nodes. Ideally, creating a distributed data collector should not be much harder than creating a [multiprocess one](https://github.com/facebookresearch/rl/blob/ac044e4a4795fa29ab38c42fb8e0ddc95ef4503d/torchrl/collectors/collectors.py#L565). The inputs to pass to this data collector for construction are the following:
- A list of environment creation functions: this may not be easy to pass from node to node. Perhaps `CloudpickleWrapper` could handle that? Otherwise we'd need to pass a set of instructions to the node on how to build the environment, which can be tedious in some instance and would require us to think carefully about the API.
- Collection parameters: total number of frames to collect, frames per batch, whether we should start with random policy, etc.
- Device on which the policy should be placed and device on which the data will be placed before it's sent to the main worker
- It would be also nice to pass env variables for the distant node

**Passing data**:
On AWS clusters, it sometimes happens that those nodes can use a common physical memory system (fsx) where shared tensord can be stored. We already have `MemmapTensors` in torchrl to handle this use case and pass data efficiently from node to node.

cc @H-Huang for suggestions

**Comment by H-Huang** [comment]
This is great to see and very exciting! Taking a step back, I am curious how do you envision the user experience is when starting up a distributed reinforcement learning job? The typical use for distributed training in PT is the user will run a script on each node to bootstrap the training. [torch.distributed.run](https://pytorch.org/docs/master/elastic/run.html?highlight=torch%20distributed%20run#module-torch.distributed.run) has been popular for this. The process looks something as follows:

1. Run `torchrun my_training_script.py` for each node in your cluster. `my_training_script.py` contains the logic that you specify, env varaibles are passed in such as`local_rank` and `global_rank` to handle different logic for each process/node.
2. Under the hood, each node will perform initialization and join the group. `my_training_script.py` does not start until the nodes have rendezvoused.
3. Once all nodes have entered the group, your training script runs, Your script contains all the logic to handle communication between nodes, whether that is using distributed data parallel, remote procedure calls, MemmapTensors, etc.

The reason I ask is that in this manner, for your scenario, arguments such as environment creation functions and passed in env variables would be previously decided by the user and the logic would be individually written for each node. The only data that needs to be communicated is runtime data during the job such as the collection parameters / device placement as you mentioned. Please correct me if my understanding of the previous two statements are wrong.

On the flip side, if we automate this process out and offer a more comprehensive solution (a single function such as the multiprocessing dataloader) where the user does not have to handle with different logics for each rank, then the API  for this would change as well because we would have to take in the IP addresses of each node in the cluster and surface connectivity issues for each node to the user.

**Comment by vmoens** [comment]
Thanks for this @H-Huang 

On the high level, I think what I would see fit for torchrl would be a data collector class that requires a minimum understanding of what is going on under the hood for the users.

Also, most of it should be contained in the class (e.g. in `torchrl/collectors/distributed_collector.py`) and as few as possible of the core parts of the distributed pipeline should be coded in the script (e.g. in `examples/ddpg/ddpg_distributed.py`) and in the trainer (e.g. `torchrl/trainers/distributed_trainer.py`). The reason for this is that examples are examples: what torchrl provides is low-level primitives that enable fast implementation of new algorithms. Hence, coding a distributed data collector should be fast (given the doc, tutorials and examples at hand obviously)!

That being said, I understand that having a code like this might be necessary and that's fine, but it should as short as possible:

```python
def worker(...):
    # collect data given passed instruction

def main(...):
    # launch other nodes
    # make a "main" data collector that collects data from distant nodes
    # build trainer and train algorithm given main data collector

```
Hope that makes sense

**Comment by vmoens** [comment]
Another point: a big advantage of RPC would be fault-tolerance (ie. using elastic RPC). 
There are many cases in RL where one relies on an external software to gather data, but the stability of those softwares is not always guaranteed. We should keep in mind that nodes should be allowed to exit and enter at will during training.

cc @AurRod
